integration/test/Test/NginxZAuthModule.hs

```diff
@@ -0,0 +1,181 @@
+module Test.NginxZAuthModule where
+
+import API.Brig
+import API.Common
+import Control.Monad.Codensity
+import Control.Monad.Reader
+import qualified Data.ByteString as BS
+import Data.List.Extra
+import Data.Streaming.Network
+import qualified Network.HTTP.Client as HTTP
+import Network.HTTP.Types
+import Network.Socket (Socket)
+import qualified Network.Socket as NS
+import qualified Network.Socket.ByteString as NSB
+import SetupHelpers
+import System.FilePath ((</>))
+import System.IO (writeFile)
+import System.IO.Temp
+import System.Posix
+import System.Process (getPid)
+import Testlib.Prelude
+import Text.RawString.QQ
+import UnliftIO (bracket)
+import UnliftIO.Async (async, waitBoth)
+import qualified UnliftIO.Async as Async
+import UnliftIO.Directory
+import UnliftIO.Process
+import UnliftIO.Timeout (timeout)
+
+testBearerToken :: (HasCallStack) => App ()
+testBearerToken = do
+  runCodensity withTestNginz $ \port -> do
+    alice <- randomUser OwnDomain def
+    email <- asString $ alice %. "email"
+    loginResp <- login alice email defPassword >>= getJSON 200
+    token <- asString $ loginResp %. "access_token"
+
+    req0 <- HTTP.parseRequest "http://localhost"
+    let req =
+          req0
+            { HTTP.port = port,
+              HTTP.requestHeaders = [(hAuthorization, fromString $ "Bearer " <> token)]
+            }
+    submit "GET" req `bindResponse` \resp -> do
+      resp.status `shouldMatchInt` 200
+      resp.json %. "user" `shouldMatch` (alice %. "qualified_id.id")
+
+testAWS4_HMAC_SHA256_token :: (HasCallStack) => App ()
+testAWS4_HMAC_SHA256_token = do
+  runCodensity withTestNginz $ \port -> do
+    alice <- randomUser OwnDomain def
+    email <- asString $ alice %. "email"
+    loginResp <- login alice email defPassword >>= getJSON 200
+    token <- asString $ loginResp %. "access_token"
+
+    req0 <- HTTP.parseRequest "http://localhost"
+
+    let mkReq authHeader =
+          req0
+            { HTTP.port = port,
+              HTTP.requestHeaders = [(hAuthorization, authHeader)]
+            }
+        headersToTest =
```

Comment:
I read "convert headers to test" here.  maybe `testCases`?  the fact that it's headers is neither surprising nor hard to infer.